### ISSUE TITLE
Dependency "hotscript" should not be just a dev-dependency in ra-ui-material-ui

### ISSUE THREAD

**Issue by VGalata** [marmelab/react-admin#8959]
**What you were expecting:**

Correct type inference for property `source` in `FieldProps`.

**What happened instead:**

Incorrect type inference for property `source` for components using `FieldProps`.

**Environment**

* React-admin version: `4.11.0`

**Comment by fzaninotto** [comment]
Hi, I don't understand your problem. If you need type inference, you're in development environment, right?

**Comment by ghost91-** [comment]
Just as a note: The errors could probably also be avoided by specifying `"skipLibCheck": true` in our `tsconfig.json` file, but there are other good reason for not doing that. And even if we did that, you would still need to install `hotscript` as a devDependency for the typing to work correctly (otherwise, you'll probably just get some `any` types).

Ideally, we would want to mark hotscript as a "peerDevDependency", i.e., a dependency that is only needed at build time, but projects using RA also need it. But unfortunately, npm does not support such a thing :/

**Comment by fzaninotto** [comment]
@tommylenz 4.11.1 has fixed the initial issue bug, so this is a new bug. Please open a new issue with a reproduction, and we'll take a look

**Comment by hakanErgin** [comment]
Following due to same issue described [here](https://github.com/marmelab/react-admin/issues/8959#issuecomment-1577240639)

**Comment by hakanErgin** [comment]
> I don't know if this hits the same issues, but I just updated code from react-admin@4 to react-admin@latest, i.e., react-admin@4.11.1, and very basic code stopped working. Example:
> 
> ```ts
> function DatasourceList() {
>   const translate = useTranslate()
>   return (
>     <List>
>       <Datagrid rowClick="show">
>         <TextField
>           source="name"
>           label={translate("res.datasource.nameLabel")}
>         />
>         <TypeLabelField />
>         <DateField
>           source="lastupdate"
>           label={translate("res.datasource.lastupdateLabel")}
>         />
>         <TextField
>           source="description"
>           label={translate("res.datasource.descriptionLabel")}
>         />
>         <EditButton />
>       </Datagrid>
>     </List>
>   )
> }
> ```
> 
> yiels the following error
> 
> ```
> ERROR in src/resources/datasources/DatasourceList.tsx:71:11
> TS2322: Type '"description"' is not assignable to type '`[${key}]` | `[${key}].${string}` | null | undefined'.
>     69 |         />
>     70 |         <TextField
>   > 71 |           source="description"
>        |           ^^^^^^
>     72 |           label={translate("res.datasource.descriptionLabel")}
>     73 |         />
>     74 |         <EditButton />
> ```
> 
> I haven't found any issue like this, but I don't know how this could be a problem on our end as I just updated react-admin and now I am running into this issue. If I can't fix this, I will try to reproduce it in some example app and will file a proper issue. For the time being I'd be happy for any advice. Thanks!

Hi, did you come up with a solution for this?

**Comment by tommylenz** [comment]
@hakanErgin No, unfortunately I didn't look into this since. But the issue still exists and maybe I have some time for this in the following days.

**Comment by fzaninotto** [comment]
Again, if you want us to look at an issue, make sure you open a new issue ticket with a simple reproduction.

**Comment by hakanErgin** [comment]
@tommylenz Thanks for your reply. The hack I found was in `node_modules/hotscript/dist/internals/objects/Objects.d.ts` line 530: `return: Impl.AllPaths<this["arg0"]>;` -> `[any](return: any;)` so far.

@fzaninotto Since I have changed nothing accept installing a newer version of RA, I do not know how to reproduce this. That's why I'm wondering if anybody else has a solution.

**Comment by tommylenz** [comment]
@hakanErgin Thanks for the tip, I will keep that in mind. Hotscript seems a bit odd for me as well, since tsc fails on hotscript's most recent version. I tried to, but wasn't able to reproduce this error with a clean react-admin project. If I can, I will obviously open a new issue @fzaninotto.

**Comment by qisaw** [comment]
Hi, I had this exact issue recently upgrading to the latest version of `react-admin` `4.16.16`

Turns out this was due to my typescript version (I had `4.X` and upgrading to `5.X` solved this)
